core/sdk-core/src/test/java/software/amazon/awssdk/core/internal/async/ByteArrayAsyncRequestBodyTest.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.async;
+
+import static org.junit.Assert.assertTrue;
+import java.nio.ByteBuffer;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.junit.Test;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+public class ByteArrayAsyncRequestBodyTest {
+    private class testSubscriber implements Subscriber<ByteBuffer> {
+        private Subscription subscription;
+        protected AtomicBoolean onCompleteCalled = new AtomicBoolean(false);
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            this.subscription = s;
+            CompletableFuture<Void> completableFuture1 = CompletableFuture.runAsync(() -> s.request(1));
```

Comment:
I don't get the purpose of this test. [2.7 of the Reactive Streams spec](https://github.com/reactive-streams/reactive-streams-jvm/blob/v1.0.2/README.md) requires external synchronization on all calls to a `Subscription`:

> A Subscriber MUST ensure that all calls on its Subscription take place from the same thread or provide for respective external synchronization.

So this is not something we need to account for.

cc: @zoewangg

Reply:
Reading @zoewangg's previous review comments, I think she had this in mind

When `onComplete()` is called, try and call `subscription.request(1)` inside of `onComplete()` and see if `onNext` is called again. Is that right Zoe?

Reply:
I didn't know this was against the spec. I don't remember how exactly http client subscriber ended up sending out concurrent demands. I thought this was the sequence:
`subscriber.request -> subscriber.request -> subscriber.onNext -> subscriber.onComplete`, but I might be wrong.
We can probably try invoke `request` in `onComplete` as you mentioned and see if it throws statckoverflowException without the fix.

Reply:
I tried calling `subscription.request` asynchronously in both `onSubscribe()` and `onComplete()`, and this would throw out a StackOverflowError without the change. But it works after the change.